### PR TITLE
Feature airfoil import bug

### PR DESCRIPTION
**Pull request by mclarke2**
Addresses #402
Reinstituting the option of no airfoil.

### CODE REVIEW

File: regression/scripts/Vehicles/Tiltwing.py [hunk 1]
```diff
@@ -260,7 +260,11 @@ def vehicle_setup():
     rot.design_altitude          = 500 * Units.feet                  
     Lift                         = vehicle.mass_properties.takeoff*9.81  
     rot.design_thrust            = (Lift * 1.5  )/net.number_of_engines  
-    rot.induced_hover_velocity   = np.sqrt(Lift/(2*rho*rot.disc_area*net.number_of_engines))                     
+    rot.induced_hover_velocity   = np.sqrt(Lift/(2*rho*rot.disc_area*net.number_of_engines))
```
Comment (planes) [inline regression/scripts/Vehicles/Tiltwing.py:263]: Not for this PR. But this equation is not very SUAVE style.

File: regression/scripts/Vehicles/X57_Maxwell.py [hunk 2]
```diff
@@ -181,119 +181,10 @@ def vehicle_setup():
...
-       
-    # Segment                                   
-    segment                                     = SUAVE.Components.Fuselages.Segment()
-    segment.tag                                 = 'segment_11'   
-    segment.percent_x_location                  = 1 
-    segment.percent_z_location                  = 0.04787575562  
-    segment.height                              = 0.1251798561 
-    segment.width                               = 0.1206021048 
-    fuselage.Segments.append(segment)             
-
```
Comment (planes) [inline regression/scripts/Vehicles/X57_Maxwell.py:294]: Why did you delete these?

File: regression/scripts/battery_propeller/battery_propeller.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def main():
     
     # save results 
     # This should be always left uncommented to test the SUAVE's Input/Output archive functions
-    save_results(results)
+    #save_results(results)
```
Comment (planes) [inline regression/scripts/battery_propeller/battery_propeller.py:44]: this needs to be uncommented, see comment

File: trunk/SUAVE/Components/Energy/Converters/Propeller.py [hunk 4]
```diff
@@ -259,12 +260,38 @@ def spin(self,conditions):
...
-                Cl[:,jj]    = cl_sur[a_geo[a_loc[jj]]](Re[:,jj],alpha[:,jj],grid=False)  
-                Cdval[:,jj] = cd_sur[a_geo[a_loc[jj]]](Re[:,jj],alpha[:,jj],grid=False) 
+            
+            # If propeller airfoils are defined, using airfoil surrogate 
+            if a_loc != None:
+                # Compute blade CL distribution from the airfoil data 
+                Cl    = np.zeros((ctrl_pts,Nr))              
+                Cdval = np.zeros((ctrl_pts,Nr)) 
+                for jj in range(Nr):                 
+                    Cl[:,jj]    = cl_sur[a_geo[a_loc[jj]]](Re[:,jj],alpha[:,jj],grid=False)
```
Comment (planes) [inline trunk/SUAVE/Components/Energy/Converters/Propeller.py:270]: This loop needs to be  removed, it's going to be soooooooo sloooooww

File: trunk/SUAVE/Components/Energy/Converters/Rotor.py [hunk 5]
```diff
@@ -263,12 +264,38 @@ def spin(self,conditions):
...
-                Cl[:,jj]    = cl_sur[a_geo[a_loc[jj]]](Re[:,jj],alpha[:,jj],grid=False)  
-                Cdval[:,jj] = cd_sur[a_geo[a_loc[jj]]](Re[:,jj],alpha[:,jj],grid=False) 
+            
+            # If rotor airfoils are defined, using airfoil surrogate 
+            if a_loc != None:
+                # Compute blade CL distribution from the airfoil data 
+                Cl    = np.zeros((ctrl_pts,Nr))              
+                Cdval = np.zeros((ctrl_pts,Nr)) 
+                for jj in range(Nr):                 
+                    Cl[:,jj]    = cl_sur[a_geo[a_loc[jj]]](Re[:,jj],alpha[:,jj],grid=False)
```
Comment (planes) [inline trunk/SUAVE/Components/Energy/Converters/Rotor.py:274]: same comment

File: trunk/SUAVE/Methods/Propulsion/propeller_design.py [hunk 6]
```diff
@@ -118,37 +118,15 @@ def propeller_design(prop,number_of_stations=20):
...
-                   path +  'NACA_4412_polar_Re_500000.txt' ,
-                   path +  'NACA_4412_polar_Re_1000000.txt' ]]   # airfoil polars for at different reynolds numbers  
-       
-        # 0 represents the first airfoil, 1 represents the second airfoil etc. 
-        a_loc = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]  
-    
-        prop.airfoil_geometry        = a_geo
-        prop.airfoil_polars          = a_pol     
-        prop.airfoil_polar_stations  = a_loc     
+        airfoil_flag    = False
```
Comment (planes) [inline trunk/SUAVE/Methods/Propulsion/propeller_design.py:123]: It may be worth putting an a print here too like before. Defaulting to a scaled DAE51.

File: trunk/SUAVE/Methods/Propulsion/propeller_design.py [hunk 7]
```diff
@@ -165,41 +143,50 @@ def propeller_design(prop,number_of_stations=20):
...
+                AoA_old_guess = 0.01
+                cl_diff       = 1  
+                broke         = False   
+                ii            = 0
                 
-            alpha[i] = AoA_old_guess     
-            Cdval[i] = airfoil_cd_surs[a_geo[a_loc[i]]](RE[i],alpha[i],grid=False)  
-
+                # Newton Raphson Iteration 
+                while cl_diff > 1E-3:
```
Comment (planes) [inline trunk/SUAVE/Methods/Propulsion/propeller_design.py:162]: This might be better behaved with a scipy function running this. 1E-3 is not a tight tolerance.

File: trunk/SUAVE/Methods/Propulsion/propeller_design.py [hunk 8]
```diff
@@ -165,41 +143,50 @@ def propeller_design(prop,number_of_stations=20):
...
+                    AoA_old_guess  = AoA_new_guess 
+                    
+                    # compute difference for tolerance check
+                    cl_diff        = abs(Cl_guess)      
+                     
+                    ii+=1 	
+                    if ii>10000:	
+                        # maximum iterations is 10000
+                        print('Propeller/Rotor section is not converging to solution')
+                        broke = True
```
Comment (planes) [inline trunk/SUAVE/Methods/Propulsion/propeller_design.py:181]: I realize you copied this from the propeller script, you don't need broke in this case since you're not saving that output for later

File: trunk/SUAVE/Methods/Propulsion/propeller_design.py [hunk 9]
```diff
@@ -282,12 +269,19 @@ def propeller_design(prop,number_of_stations=20):
...
-    airfoil_geometry_data = import_airfoil_geometry(a_geo)
-    for i in range(N):
-        t_c[i]   = airfoil_geometry_data.thickness_to_chord[a_loc[i]]    
-        t_max[i] = airfoil_geometry_data.max_thickness[a_loc[i]]*c[i]
-        
+    t_c    = np.zeros(N)   
+    if airfoil_flag:     
+        airfoil_geometry_data = import_airfoil_geometry(a_geo)
+        for i in range(N):
+            t_max[i] = airfoil_geometry_data.max_thickness[a_loc[i]]*c[i]
```
Comment (planes) [inline trunk/SUAVE/Methods/Propulsion/propeller_design.py:276]: Get rid of the for loops

File: regression/scripts/Vehicles/Tiltwing.py (hunk 1 above)
Comment (mclarke2) [reply regression/scripts/Vehicles/Tiltwing.py:263]: I agree.

File: regression/scripts/Vehicles/X57_Maxwell.py (hunk 2 above)
Comment (mclarke2) [reply regression/scripts/Vehicles/X57_Maxwell.py:294]: Opps. That was a mistake. Replaced

File: trunk/SUAVE/Methods/Propulsion/propeller_design.py (hunk 6 above)
Comment (mclarke2) [reply trunk/SUAVE/Methods/Propulsion/propeller_design.py:123]: Comment placed

File: trunk/SUAVE/Methods/Propulsion/propeller_design.py (hunk 8 above)
Comment (mclarke2) [reply trunk/SUAVE/Methods/Propulsion/propeller_design.py:181]: Got it, removed.

File: trunk/SUAVE/Methods/Propulsion/propeller_design.py (hunk 9 above)
Comment (mclarke2) [reply trunk/SUAVE/Methods/Propulsion/propeller_design.py:276]: loops removed

File: trunk/SUAVE/Methods/Propulsion/propeller_design.py (hunk 7 above)
Comment (mclarke2) [reply trunk/SUAVE/Methods/Propulsion/propeller_design.py:162]: scipy function used

File: trunk/SUAVE/Components/Energy/Converters/Propeller.py (hunk 4 above)
Comment (mclarke2) [reply trunk/SUAVE/Components/Energy/Converters/Propeller.py]: loop removed

File: trunk/SUAVE/Components/Energy/Converters/Rotor.py (hunk 5 above)
Comment (mclarke2) [reply trunk/SUAVE/Components/Energy/Converters/Rotor.py]: addressed. see above comment

File: trunk/SUAVE/Methods/Propulsion/electric_motor_sizing.py [hunk 10]
```diff
@@ -179,25 +179,45 @@ def optimize_kv(io, v , omeg,  etam ,  Q, kv_lower_bound =  0.01, Res_lower_boun
...
-    cons2 = [{'type':'eq', 'fun': constraint_2,'args': args}] 
+    slack_cons = [{'type':'eq', 'fun': slack_constraint_1,'args': args},
+                  {'type':'eq', 'fun': slack_constraint_2,'args': args}] 
+  
+    torque_con = [{'type':'eq', 'fun': hard_constraint_2,'args': args}]
     
     bnds = ((kv_lower_bound, kv_upper_bound), (Res_lower_bound , Res_upper_bound))
-    sol = minimize(objective, [0.5, 0.1], args=(v , omeg,  etam , Q , io) , method='SLSQP', bounds=bnds, tol=1e-6, constraints=cons1) 
+    
+    # try hard constraints to fine optimum motor parameters
```
Comment (planes) [inline trunk/SUAVE/Methods/Propulsion/electric_motor_sizing.py:192]: fine or find?

File: trunk/SUAVE/Methods/Propulsion/propeller_design.py [hunk 11]
```diff
@@ -165,41 +145,25 @@ def propeller_design(prop,number_of_stations=20):
...
+        if airfoil_flag:   
+            # compute airfoil polars for airfoils 
+            airfoil_polars  = compute_airfoil_polars(a_geo, a_pol)  
+            airfoil_cl_surs = airfoil_polars.lift_coefficient_surrogates 
+            airfoil_cd_surs = airfoil_polars.drag_coefficient_surrogates 
+            
+            alpha = np.zeros_like(RE)
+            Cdval = np.zeros_like(RE)  
+            
+            for i in range(N):
```
Comment (planes) [inline trunk/SUAVE/Methods/Propulsion/propeller_design.py:158]: Does this need to be in a for loop? A newton solver should be able to do this is as a vector.

**Review by planes — APPROVED** [review]
See comments. It's pretty much done.

File: trunk/SUAVE/Methods/Propulsion/electric_motor_sizing.py (hunk 10 above)
Comment (mclarke2) [reply trunk/SUAVE/Methods/Propulsion/electric_motor_sizing.py:192]: find. will change

**Comment by planes** [comment]
Dang this PR looks good! I want to approve it again.
